File: Data Structures & Algorithms/two-integer-sum/submission-4.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        s = {}
        for i in range(len(nums)):
            s[nums[i]] = i

        print(s)
        l = []

        for i in range(len(nums)):
            if target - nums[i] in s and s[target - nums[i]] != i:
                l.append(i)
                l.append(s[target-nums[i]])
                return l



        

File: Data Structures & Algorithms/combination-target-sum/submission-4.cpp
class Solution {
public:
    vector<vector<int>> res;
    void backtracking(int index, int target, vector<int>& path, int currSum, vector<int>& nums) {


        if (currSum > target || index >= nums.size()) {
            return;
        } 

        if (currSum == target) {
            res.push_back(path);
            return;
        }

        //choice 1: add number and stay on the index
        path.push_back(nums[index]);
        backtracking(index,target,path,currSum + nums[index], nums);
        path.pop_back();

        //choice 2: move on to next index
        backtracking(index + 1,target,path,currSum, nums);

    }
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> p = {};
        backtracking(0,target,p,0,nums);

        return res;
    }
};
